Documentation for WFST decoding (#161)
nvidia-riva/tutorials: wfst_class_lm.md

modifying `G.fst` to support paths in the acceptor that refers to word classes.
This requires that
* The lexicon has all entries from both the language model and the entries in
each of word classes. This is important as otherwise the words in the word 
classes will not be recognized.
* The n-gram model should have the relevant phrases build from word class grammars.
For example in the above example the word class grammar would 
be *"I would like to fly from **#entity:city** to **#entity:city**"* 
with **#entity:city**
being the label for the word class corresponding to cities with airports. 
This can be achieved in the following two ways:
    - Build the language model from text containing standard text as well as
  word class labels
    - n-gram LM interpolation/mixing of two LMs, one from standard text and other 
  from domain specific grammars.

The latter option provides for better weighting between in-domain and out of 
domain LMs. 
Please refer to the tutorial ["How To Train, Evaluate, and Fine-Tune an n-gram Language Model"](https://github.com/nvidia-riva/tutorials/blob/main/asr-python-advanced-nemo-ngram-training-and-finetuning.ipynb) 

> for LM model interpolation using kenlm. 

Reviewer: ```suggestion
for LM model interpolation using KenLM. 
```

> Alternatively [SRILM](http://www.speech.sri.com/projects/srilm/) (license required) can also be used here

Reviewer: ```suggestion
Alternatively, [SRILM](http://www.speech.sri.com/projects/srilm/) (license required) can also be used here.
```

The process for supporting word classes in the WFST framework basically involves
replacement of the arcs representing the class labels, with FSTs created from 
the entities present in the word classes. Given a `G.fst` containing 
word class labels (`#entity:<class>`), and the files (filenames `<class>.txt`) 

> containing the list of entities for the class. The list of entities  is not 

Reviewer: ```suggestion
containing the list of entities for the class. The list of entities is not 
```

> restricted to words and could be phrases

Reviewer: ```suggestion
restricted to words and could be phrases.
```

> For example, in the above example `cities.txt` would contain at the very least the 

Reviewer: ```suggestion
For example, in the above example, `cities.txt` would contain at the very least the 
```

following
```shell
$> cat city.txt
london
new york
paris
seattle
```

> ## Process to create the `TLG.fst` supporting word classes

Reviewer: ```suggestion
## Process to Create `TLG.fst` Supporting Word Classes
```

Checkout the `class_lm` branch from https://github.com/anand-nv/riva-asrlib-decoder/tree/class_lm. 

> All the below scripts are available in `src/riva/decoder/scripts/prepare_TLG_fst` path.

Reviewer: ```suggestion
The following scripts are available in the `src/riva/decoder/scripts/prepare_TLG_fst` directory path.
```

> * Generate the lexicon, list of Tokens/units and list of entities from the 

Reviewer: ```suggestion
* Generate the lexicon, the list of tokens/units, and the list of entities from the 
```

> arpa file. This is accomplished using `lexicon_from_arpa.py` script. 

Reviewer: ```suggestion
arpa file. This is accomplished using the `lexicon_from_arpa.py` script. 
```

The syntax for the command is as follows:

`lexicon_from_arpa.py --asr_model <path_to_nemo_model> --lm_path <path_to_arpa_file> --grammars <path_to_folder_containing_grammars> <units_txt_path:output> <lexicon_path:output> <classlabels_path:output>`

> This generates the list of units(tokens), the lexicon and the list of word 

Reviewer: ```suggestion
This generates the list of units(tokens), the lexicon, and the list of word 
```

> classes (to be used as disambiguation symbols) 

Reviewer: ```suggestion
classes (to be used as disambiguation symbols). 
```

* Generate the FSTs, `T.fst`, `L.fst` and `G.fst`, extend and compose to 

> form `TLG.fst`

Reviewer: ```suggestion
form `TLG.fst`.
```

`mkgraph_ctc_classlm.sh --units <units_txt_path> --extra_word_disamb <classlabels_path> <lexicon_path>  <path_to_arpa_file> <lang_folder_path:output>`

When run successfully, the above command will generate a graph folder in

> `<lang_folder_path>` with the name `graph_CTC_<lm_name>` which contains

Reviewer: ```suggestion
`<lang_folder_path>` with the name `graph_CTC_<lm_name>`, which contains
```

> all the files relevant for decoding with WFST

Reviewer: ```suggestion
all the files relevant for decoding with WFST.
```

## Evaluation
The `evaluate_nemo_wfst.py` script can help evaluate the WFST decoder before 

> being used in the Riva service maker build. This would be useful to 

Reviewer: ```suggestion
being used in the Riva ServiceMaker build. This would be useful to 
```

debug any issues with `TLG.fst`

> The script can be run as follows

Reviewer: ```suggestion
The script can be run as follows:
```

`./evaluate_nemo_wfst.py <model_path> <graph_path> <path_containing_audio_files> <results_txt_file:output>`

## Deploying in Riva

> The generated `TLG.fst` can be used with the kaldi decoder in riva. To build the riva ASR service using the generated

Reviewer: ```suggestion
The generated `TLG.fst` can be used with the Kaldi decoder in Riva. To build the Riva ASR service using the generated
```

> `TLG.fst` the following command can be used

Reviewer: ```suggestion
`TLG.fst`, run:
```
